rp2040-hal/src/pll.rs

```diff
@@ -0,0 +1,227 @@
+//! Phase-Locked Loops (PLL)
+// See [Chapter 2 Section 18](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+
+use core::{
+    convert::{
+        Infallible,
+        TryFrom
+    },
+    marker::PhantomData,
+    ops::{
+        RangeInclusive,
+        Range,
+        Deref
+    }
+};
+
+use embedded_time::{
+    fixed_point::FixedPoint,
+    rate::{
+        Hertz,
+        Generic
+    }
+};
+
+use nb::Error::WouldBlock;
+
+/// State of the PLL
+pub trait State {}
+
+
+/// PLL is disabled but is configured.
+pub struct Disabled {
+    refdiv: u8,
+    vco_freq: Hertz,
+    post_div1: u8,
+    post_div2: u8
+}
+
+/// PLL is locked : it delivers a steady frequency.
+pub struct Locked;
+
+/// PLL is locking into its designated frequency.
+pub struct Locking {
+    post_div1: u8,
+    post_div2: u8
+}
+
+impl State for Disabled {}
+impl State for Locked {}
+impl State for Locking {}
+
+
+/// Trait to handle both underlying devices from the PAC (PLL_SYS & PLL_USB)
+pub trait PhaseLockedLoopDevice: Deref<Target = rp2040_pac::pll_sys::RegisterBlock> {}
+
+impl PhaseLockedLoopDevice for rp2040_pac::PLL_SYS {}
+impl PhaseLockedLoopDevice for rp2040_pac::PLL_USB {}
+
+/// A PLL.
+pub struct PhaseLockedLoop<S: State, D: PhaseLockedLoopDevice> {
+    device: D,
+    state: S
+}
+
+impl<S: State, D: PhaseLockedLoopDevice> PhaseLockedLoop<S, D> {
+    fn transition<To: State>(self, state: To) -> PhaseLockedLoop<To, D> {
+        PhaseLockedLoop {
+            device: self.device,
+            state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> D{
+        self.device
+    }
+}
+
+
+/// Error type for the PLL module.
+/// See Chapter 2, Section 18 §2 for details on constraints triggering these errors.
+pub enum Error {
+
+    /// Proposed VCO frequency is out of range.
+    VCOFreqOutOfRange,
+
+    /// Feedback Divider value is out of range.
+    FBDIVOutOfRange,
+
+    /// Post Divider value is out of range.
+    PostDivOutOfRage,
+
+    /// Reference Frequency is out of range.
+    RefFreqOutOfRange,
+
+    /// Bad argument : overflows, bad conversion, ...
+    BadArgument
+}
+
+impl<D: PhaseLockedLoopDevice> PhaseLockedLoop<Disabled, D> {
+
+
+    /// Instantiates and configures a new Phase-Locked-Loop device.
+    pub fn new(dev: D, refdiv: u8, vco_freq: Generic<u32>, post_div1: u8, post_div2: u8) -> Result<PhaseLockedLoop<Disabled, D>, Error> {
```

Comment:
I would merge the calculations in `initialize` and `new` into `initialize`. That way you could reconfigure the PLL later without using `free`. Also then you don't have to pass state between `new` and `initialize`.

Also having all the parameters in a struct would allow us to add a `calculate_params` function that calculates a fitting set of parameters for a given XOsc and desired output frequency. Splitting out the calculations from the configuration should also make it easier to test the module.

Reply:
Done, tell we what you think :)